mlx/ops.cpp

```diff
@@ -2621,7 +2621,8 @@ array quantized_matmul(
     int group_size /* = 64 */,
     int bits /* = 4 */,
     StreamOrDevice s /* = {} */) {
-  auto x = in_x;
+  auto dtype = promote_types(in_x.dtype(), scales.dtype());
+  auto x = astype(in_x, dtype, s);
```

Comment:
I remember this bug.

Should you also promote with `biases`?

Reply:
Hmm interesting. I cast everything to dtype so there wouldn't be a bug anymore but it may be better to promote with both. Let me know. Also this means I 'll end up adding the variadic template of promote_types which I don't know why we don't have tbh. Initially I wrote `promote_types(in_x.dtype(), scales.dtype(), biases.dtype())` but that doesn't exist :-) 

Ideally we cast `x` to `scales.dtype` before the call to QuantizedLinear so that we are copying the smallest of the tensors (at least during inference).

Reply:
We have a `result_type({a, b, c})` which takes arrays as input and returns the promoted type. It would be nice if it was variadic to avoid the vector construction :).

It's directly inspired from NumPY https://numpy.org/doc/stable/reference/generated/numpy.result_type.html

Reply:
Oh nice, shame on me for not finding that. I can add a recursive variadic version of it if you want, don't know how much we 'd be saving but why not?

Reply:
I think a couple of reasonable options are:
1. Use `result_type({x, s, b})` (which is consistent with all of our other ops so that's nice
2. Always cast `x` to the type of `s` and `b` and enforce that `s.dtype() == b.dtype()`? 

In the bug I had before, I was sending in a 32-bit `x` and `s` and `b` were `float16`. So what I wanted there was actually for `x` to be `float16` like `s` and `b` (more like option 2).

Reply:
> I can add a recursive variadic version of it if you want, don't know how much we 'd be saving but why not?

Lol yea I wouldn't do it for perf reasons.. I'm cool either way... very fine with just using the vector version.